Fix a segfault in RocksDB iterator

jiacheliu3 opened the pull request:
### What changes are proposed in this pull request?

This change intends to fix RocksDB segfaults caused by race conditions when we close the RocksDB concurrent to ongoing operations like:
1. Iterations represented by BlockMaster.validateBlocks()
2. Backup/checkpoint where we iterate the RocksDB contents and dump into an output stream
3. RPC serving in RocksInodeStore

The solution is:
1. Use an ReadWriteLock to control RocksDB access. R/W will take the read lock while shutdown/restart takes write lock.
2. Add a lock-less flag which indicates ongoing R/W should abort early instead of finishing, so the write lock does not starve.

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:
```diff
@@ -442,6 +471,9 @@ public Long next() {
   @Override
   public Optional<Inode> getChild(Long inodeId, String name, ReadOption option) {
+    if (mClosed.get()) {
+      throw new RuntimeException("RocksDB is closed. Master is failing over or shutting down.");
```
Honestly I'm not sure this is the best exception to throw. Feel free to comment if you got a better one :)

tcrain replied:
Maybe alluxio.exception.runtime.UnavailableRuntimeException

tcrain commented:
Do you think with this fix, we could add something like in this PR?
https://github.com/Alluxio/alluxio/pull/15657

I remember we never added that PR because we weren't sure if RocksDB was being accessed safely, but with this fix, maybe we can do both?

elega commented on core/common/src/main/java/alluxio/conf/PropertyKey.java:
```diff
@@ -782,6 +782,13 @@ public String toString() {
               + " be used.", Name.CONF_DIR))
           .setScope(Scope.ALL)
           .build();
+  public static final PropertyKey ROCKS_GRACEFUL_SHUTDOWN_TIMEOUT =
```
i guess based on our discussion we don't actually need this correct?

jiacheliu3 replied:
yes if in our review we think the flag + RWLock approach is handling all race conditions, I will remove the grace period and all relevant code

tcrain replied:
So I think you can keep this, but just wait for a maximum amount of time, and if you can't shut down within this time, you log a warning and say that we are shutting down unsafely because a thread still has a lock on the object.

Basically if we see this warning log, we know something is not working correctly that we need to debug.

elega commented on core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:
```diff
@@ -99,30 +103,63 @@ public static long readLong(byte[] bytes, int start) {
+          rocksIterator.next();
+          succeeded = true;
+          return result;
         } catch (Exception exc) {
```
can we just catch throwable? it should cover exception

jiacheliu3 replied:
the existing behavior is wrapping `Exception` into `RuntimeException`, i didn't want to change that

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:
```diff
@@ -600,10 +652,12 @@ private RocksDB db() {
+         RocksIterator inodeIter = db().newIterator(mInodesColumn.get(), readOptions)) {
       inodeIter.seekToFirst();
       while (inodeIter.isValid()) {
+        abortIfClosing();
```
this is the tricky part...

`abortIfClosing()` checks the `mClosed` flag, which is set BEFORE acquiring the lock. Therefore we can abort here successfully. See the `close()` method implementation in this class.

jiacheliu3 commented on core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:
```diff
@@ -99,30 +103,63 @@ public static long readLong(byte[] bytes, int start) {
-            rocksIterator.close();
+          if (!succeeded) {
             valid.set(false);
+            rocksIterator.close();
```
An iterator is never intended for concurrent use, so I don't think it's used like that in Alluxio. I don't think `Iterator` provides thread safety. If what you describe really occurs, we should fix that concurrent usage instead :P

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java:
```diff
@@ -378,21 +405,49 @@ public void addLocation(long id, BlockLocation location) {
+  }
+
+  // TODO(jiacheng): double check what happens if max lock count error here
+  public LockResource checkAndAcquireReadLock() {
```
Currently this is what I'm not sure about. 

On a max lock count error, a `java.lang.Error` will be thrown and the thread will die. But most of these calls are from RPC threads and dead RPC threads don't crash the process. What's the best thing to do here? If I just catch Error and `ProcessUtils.fatalError()`, would that be an overkill?

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -444,6 +444,10 @@ public JournalEntry next() {
         if (!hasNext()) {
           throw new NoSuchElementException();
         }
+        /*
+         * When the BlockStore is RocksBlockMetaStore, the iterator takes care of thread safety
+         * in hasNext() and next() calls.
+         */
```
i will add comment like this on 2 places:
1. RocksUtils where the iter is created
2. Where this iter is used

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java:
```diff
@@ -79,6 +85,20 @@ public class RocksBlockMetaStore implements BlockMetaStore {
   private final AtomicReference<ColumnFamilyHandle> mBlockMetaColumn = new AtomicReference<>();
   private final AtomicReference<ColumnFamilyHandle> mBlockLocationsColumn = new AtomicReference<>();
   private final LongAdder mSize = new LongAdder();
+  /**
+   * When this is true, stop serving all requests as the RocksDB is being closed.
+   * This flag serves two purposes:
+   * 1. Before an r/w operation, if this is observed then bail the operation early.
+   * 2. During an r/w operation like iteration, if this is observed then abort the iteration.
+   */
+  private final AtomicBoolean mClosed = new AtomicBoolean(false);
+  /**
+   * The RWLock is used to guarantee thread safety between r/w and RocksDB closing/restarting.
+   * When the RocksDB is closing/restarting, no concurrent r/w should be present.
+   * Therefore, we acquire the write lock on closing/restarting, and acquire a read lock for
+   * reading/writing the RocksDB.
+   */
+  private final ReadWriteLock mDbStateLock = new ReentrantReadWriteLock();
```
These two come in pairs. So i think i can just wrap them into an object like `RocksProtocol`? Then
```
abstract class RocksProtocol {
  final AtomicBoolean mClosed;
  final ReadWriteLock mDbStateLock;
  abstract LockResource checkAndAcquireReadLock();
  abstract void abortCheck();
}

class RocksBlockMetaStore implements BlockMetaStore extends RocksProtocol {
...
}
```

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java:
```diff
@@ -378,21 +426,47 @@ public void addLocation(long id, BlockLocation location) {
+
+  // TODO(jiacheng): double check what happens if max lock count error here
+  private LockResource checkAndAcquireReadLock() {
+    // Check before locking so if the RocksDB will be closed, abort early
```
This is what I came up for absolute correctness...
```
1. check
2. readLock
3. check again
```
I think it's hard to get rid of either check because when you grab the read lock, it may be before or after a writer.

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java:
```diff
@@ -318,8 +338,17 @@ public void removeBlock(long id) {
+    }
+    // Reset the DB state and prepare to serve again
+    LOG.info("RocksDB ready to serve again");
+    mClosed.set(false);
```
TODO: even though the DB is up again, previous iterators do not make sense anymore...

jiacheliu3 commented:
@yuzhu @tcrain Although I'm not entirely done with the corner cases I added comments about, I believe this reflects the solution we agreed on. PTAL, thanks!

jiacheliu3 commented on core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:
```diff
@@ -209,11 +219,22 @@ public void delete(PageId pageId) throws PageNotFoundException {
+      LOG.info("Closing RocksPageStore and recycling all RocksDB JNI objects");
+      if (mDb != null) {
+        try {
+          // Column handles must be closed before closing the db, or an exception gets thrown.
```
these are copied from RocksBlockMetaStore

jiacheliu3 commented on core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:
```diff
@@ -94,35 +99,60 @@ public static long readLong(byte[] bytes, int start) {
-          valid.set(false);
           throw new RuntimeException(exc);
+        } catch (Throwable t) {
+          LOG.warn("Iteration aborted because of error", t);
```
if this log is not that necessary, we can just remove this catch, assuming the uncaught ex will be logged later anyway

jiacheliu3 commented on core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:
```diff
@@ -94,35 +99,60 @@ public static long readLong(byte[] bytes, int start) {
+         * RocksDB and all relevant references without worrying about concurrent readers
+         * like this escaped iterator.
+         */
+        return (!abortCheck.get()) && valid.get() && rocksIterator.isValid();
```
if i parse in hasNext(), does that make atomicity easier?

tcrain commented:
Currently when you create an iterator over the db, you retake the lock on each access.

Maybe it would be simpler to just take the lock once on iterator creation, then release the lock when the iterator is closed.
Then on each call to hasNext() or next(), first you check mClosed, then throw an exception if closed.

The only worry here is if someone is holding an iterator while waiting or being blocked for a long time while not using the iterator, but this is probably bad behavior anyway as rocksDB iterators are not meant to be left unused.

jiacheliu3 commented:
> Currently when you create an iterator over the db, you retake the lock on each access.
> 
> Maybe it would be simpler to just take the lock once on iterator creation, then release the lock when the iterator is closed. Then on each call to hasNext() or next(), first you check mClosed, then throw an exception if closed.
> 
> The only worry here is if someone is holding an iterator while waiting or being blocked for a long time while not using the iterator, but this is probably bad behavior anyway as rocksDB iterators are not meant to be left unused.

Makes sense, I hate grabbing the lock on every `next()` too. Agreed on your 2nd point, where that worry is unlikely to happen and should not happen.

tcrain commented:
I think we should try to clean up the existing concurrency control for the rocks db objects.

I think so far you have fixed the existing concurrency control by adding another synchronization method on top. I would rather we just have 1 synchronization primitive.

Notice how currently the RocksStore.java functions are wrapped in synchronized blocks, then afterwards the state lock is added afterwards.

I think we should remove the synchronized blocks and instead of putting the state lock in a separate class, put it directly in the RocksStore class and use that for concurrency control.

Now when RocksStore.getDB() is called, instead of returning the unprotected object, it will return the object wrapped in a closable object that has the read lock taken on the state lock.

The other functions within RocksStore that are currently synchronized can instead just take the state lock as a write lock.

jiacheliu3 commented:
> I think we should try to clean up the existing concurrency control for the rocks db objects.
> 
> I think so far you have fixed the existing concurrency control by adding another synchronization method on top. I would rather we just have 1 synchronization primitive.
> 
> Notice how currently the RocksStore.java functions are wrapped in synchronized blocks, then afterwards the state lock is added afterwards.
> 
> I think we should remove the synchronized blocks and instead of putting the state lock in a separate class, put it directly in the RocksStore class and use that for concurrency control.

Thanks for the suggestions they make sense to me! I moved the locking logic into `RocksStore` and removed the old `synchronized` mechanism. Using one unified RWL set makes much more sense. I will add some benchmarking stats in a day or so.

> Now when RocksStore.getDB() is called, instead of returning the unprotected object, it will return the object wrapped in a closable object that has the read lock taken on the state lock.
>
> The other functions within RocksStore that are currently synchronized can instead just take the state lock as a write lock.

The write lock part is done. After some thoughts I didn't make `RocksStore.getDB()` return a lock. Because together with the db object, there are a few other `RocksObject` which are guarded by the same RWL (but are accessed separately to the db), like
```
private class RocksWriteBatch {
    @Override
    public void writeInode(MutableInode<?> inode) {
      try (LockResource lock = mRocksStore.checkAndAcquireReadLock()) {
        mBatch.put(mInodesColumn.get(), Longs.toByteArray(inode.getId()),
            inode.toProto().toByteArray());
      }
```

jiacheliu3 commented on core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:
```diff
@@ -111,23 +127,34 @@ public boolean hasNext() {
+        boolean succeeded = false;
+
+        // If the RocksDB wants to stop, abort the loop instead of finishing it
+        abortCheck.get();
```
if an ex is thrown here, need to close the iterator

jiacheliu3 commented on core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStoreDir.java:
```diff
@@ -66,8 +66,11 @@ public void reset() throws IOException {
 
   @Override
   public void scanPages(Consumer<Optional<PageInfo>> pageInfoConsumer) {
+    // Fix thread safety on this iterator or or demise RocksPageStore
```
talked to @dbw9580 I think we are going to demise RocksPageStore soon so won't waste time fixing it. This issue is not introduced by this change of mine.

jiacheliu3 commented:
closed because a better implementation is in https://github.com/Alluxio/alluxio/pull/17171
